blocks/hooks/custom-styles.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+import { addFilter } from '@wordpress/hooks';
+
+/**
+ * Internal dependencies
+ */
+import { getBlockRandomAnchor, hasBlockSupport } from '../api';
+
+const CUSTOM_STYLE_ATTRIBUTES = [ 'backgroundColor', 'textColor' ];
+
+const hasCustomStyleAttributes = ( attributes ) => (
+	CUSTOM_STYLE_ATTRIBUTES.some( attr => attributes[ attr ] )
+);
+
+function withCustomStylesAnchor( BlockEdit ) {
+	return class CustomStylesAnchor extends Component {
+		componentWillReceiveProps( nextProps ) {
+			if ( hasBlockSupport( nextProps.name, 'anchor' ) &&
+				hasBlockSupport( nextProps.name, 'customStyles' ) &&
+				! nextProps.attributes.anchor &&
+				hasCustomStyleAttributes( nextProps.attributes )
+			) {
+				nextProps.setAttributes( {
```

Comment:
Can you explain what `setAttributes` is doing here? In general `props` should be readonly, but it's hard to find out if this call updates props or not. Given that we check whether `nextProps.attributes.anchor` is empty, I would assume this code updates this attribute. If I correctly assumed, we might want to introduce internal state here, and combine props and state when passing down to the `<BlockEdit { ...this.props } />` in render method.

Reply:
Hi @gziolo, what we are doing here is verifying if no anchor was set yet ( by the user or automatically). If that's the case we check if attributes needing custom styles were used and if yes we set an automatically generated anchor attribute. I think it's not enough to pass a prop we really want to save the attribute for it to be serialized.

The general question here is what is our extensibility solution to allow to set an attribute depending on other. E.g: If I want to have a plugin that sets red text color attribute if the background is blue what would be the best way to do it? Right now it looks like using our hooks we can achieve this result using this solution, but maybe there is a simpler way.

Reply:
Okey, `setAttributes` is some magic functions passed down to `edit` which does all magic. It must be good 👍